// codelyzer/noHostMetadataPropertyRule.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var metadataPropertyBase_1 = require("./metadataPropertyBase");
var utils_1 = require("./util/utils");
var METADATA_PROPERTY_NAME = 'host';
var STYLE_GUIDE_LINK = 'https://angular.io/styleguide#style-06-03';
var Rule = (function (_super) {
    __extends(Rule, _super);
    function Rule(options) {
        return _super.call(this, {
            errorMessage: Rule.FAILURE_STRING,
            propertyName: METADATA_PROPERTY_NAME
        }, options) || this;
    }
    Rule.metadata = {
        description: "Disallows usage of the `" + METADATA_PROPERTY_NAME + "` metadata property.",
        descriptionDetails: "See more at " + STYLE_GUIDE_LINK + ".",
        options: null,
        optionsDescription: 'Not configurable.',
        rationale: "If you ever need to rename the property associated with @" + utils_1.Decorators.HostBinding + " or the method associated with\n    @" + utils_1.Decorators.HostListener + ", you can modify it in a single place.",
        ruleName: 'no-host-metadata-property',
        type: 'style',
        typescriptOnly: true
    };
    Rule.FAILURE_STRING = "Use @" + utils_1.Decorators.HostBinding + " or @" + utils_1.Decorators.HostListener + " rather than the `" + METADATA_PROPERTY_NAME + "` metadata property (" + STYLE_GUIDE_LINK + ")";
    return Rule;
}(metadataPropertyBase_1.MetadataPropertyBase));
exports.Rule = Rule;
