// codelyzer/usePipeTransformInterfaceRule.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var sprintf_js_1 = require("sprintf-js");
var rules_1 = require("tslint/lib/rules");
var typescript_1 = require("typescript/lib/typescript");
var utils_1 = require("./util/utils");
var PIPE_TRANSFORM = 'PipeTransform';
var Rule = (function (_super) {
    __extends(Rule, _super);
    function Rule() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Rule.prototype.apply = function (sourceFile) {
        return this.applyWithFunction(sourceFile, walk);
    };
    Rule.metadata = {
        description: "Ensures tht classes decorated with @" + utils_1.MetadataTypes.Pipe + " implement " + PIPE_TRANSFORM + " interface.",
        options: null,
        optionsDescription: 'Not configurable.',
        rationale: 'Interfaces prescribe typed method signatures. Use those signatures to flag spelling and syntax mistakes.',
        ruleName: 'use-pipe-transform-interface',
        type: 'maintainability',
        typescriptOnly: true
    };
    Rule.FAILURE_STRING = "Classes decorated with @" + utils_1.MetadataTypes.Pipe + " decorator should implement " + PIPE_TRANSFORM + " interface";
    return Rule;
}(rules_1.AbstractRule));
exports.Rule = Rule;
var hasPipeDecorator = function (node) { return !!utils_1.getDecorator(node, utils_1.MetadataTypes.Pipe); };
var hasPipeTransformInterface = function (node) { return !!utils_1.getDeclaredInterfaceName(node, PIPE_TRANSFORM); };
var validateClassDeclaration = function (context, node) {
    if (!hasPipeDecorator(node) || hasPipeTransformInterface(node))
        return;
    context.addFailureAtNode(node, sprintf_js_1.sprintf(Rule.FAILURE_STRING));
};
var walk = function (context) {
    var sourceFile = context.sourceFile;
    var callback = function (node) {
        if (typescript_1.isClassDeclaration(node))
            validateClassDeclaration(context, node);
        typescript_1.forEachChild(node, callback);
    };
    typescript_1.forEachChild(sourceFile, callback);
};
